### ISSUE TITLE
Horovod timelines are not consistent about HorovodAllreduce 

### ISSUE THREAD

**Issue by yozhao** [horovod/horovod#1912]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet) Tensor flow
2. Framework version:  1.12
3. Horovod version:  0.19.1
4. MPI version:  4.0.3
5. CUDA version: 10.0
6. NCCL version:  2.4.8
7. Python version: 3.6.9
8. OS and version: CentOS Linux release 7.4.1708
9. GCC version: gcc-5.3

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**
Please ask your question here.

I  run 2 horovod process with 2 GPUs. When I look into the timelines of one mini-batch.  I find the pattern of the 2 processes are obviously different about HorovodAllreduce.   I think HorovodAllreduce is the sync operation of the 2 process, the timeline should be similar.

The result is:
1. One timeline has large HorovodAllreduce timeline blocks, and the last HorovodAllreduce ends at about 300ms.

2.The other timeline has many and small HorovodAllreduce timeline  blocks, and the last  ends at about 210ms.

I am wondering if it is expected and what's the reason?  And 150ms+ HorovodAllreduce in the first timeline is too long. 

Thanks!

The timeline images are attached.

![hvd-tl1](https://user-images.githubusercontent.com/2783399/80354387-41063d80-88a9-11ea-92ba-721d9dae695b.jpg)

![hvd-tl2](https://user-images.githubusercontent.com/2783399/80354635-92aec800-88a9-11ea-8c62-89cbc06a25bd.jpg)

**Comment by tgaddair** [comment]
Hey @yozhao, can you explain a little bit more about the experiment setup?  The timeline is usually only generated on worker 0, so I'm not sure what you did to generate two timelines for the same process.  If you're running two separate `horovodrun` jobs in parallel, that would partially explain the disparity, as you have two processes competing for the same GPU resources.

I also don't see any NCCL messages in the timeline.  Can you run `horovodrun --check-build` to verify that NCCL support is installed?

**Comment by yozhao** [comment]
"The timeline is usually only generated on worker 0".  
Oops, I  dump timeline in each process like:

        profile_hook = tf.train.ProfilerHook(save_steps=500, output_dir='./timeline_json_{}/'.format(pid), show_memory=True)
        training_hooks = hooks
        training_hooks.append(profile_hook)
        return tf.estimator.EstimatorSpec(mode=mode, loss=total_loss, train_op=tf.group(train_op_list),
                                          training_hooks=training_hooks)


so each process will dump timeline to different folders.

**Comment by yozhao** [comment]
Hi  Travis，
In tensorflow timeline I can grep some nccl message like:  

timeline-57780-774.json-        {
timeline-57780-774.json-            "ph": "X",
timeline-57780-774.json-            "cat": "Op",
timeline-57780-774.json-            "name": "unknown",
timeline-57780-774.json-            "pid": 13,
timeline-57780-774.json-            "tid": 0,
timeline-57780-774.json-            "ts": 1587919108567382,
timeline-57780-774.json-            "dur": 25,
timeline-57780-774.json-            "args": {
timeline-57780-774.json:                "name": "_Z33ncclAllReduceRingLLKernel_sum_f328ncclColl",
timeline-57780-774.json-                "op": "unknown"
timeline-57780-774.json-            }
timeline-57780-774.json-        },

the duration is short.

Maybe you mean hvd specified timeline not tensorflow timeline?

**Comment by yozhao** [comment]
> Hey @yozhao, can you explain a little bit more about the experiment setup? The timeline is usually only generated on worker 0, so I'm not sure what you did to generate two timelines for the same process. If you're running two separate `horovodrun` jobs in parallel, that would partially explain the disparity, as you have two processes competing for the same GPU resources.
> 
> I also don't see any NCCL messages in the timeline. Can you run `horovodrun --check-build` to verify that NCCL support is installed?


Horovod v0.19.1:

Available Frameworks:
    [X] TensorFlow
    [ ] PyTorch
    [ ] MXNet

Available Controllers:
    [X] MPI
    [ ] Gloo

Available Tensor Operations:
    [X] NCCL
    [ ] DDL
    [ ] CCL
    [X] MPI
    [ ] Gloo

We have NCCL supported but no Gloo, does this matter?

**Comment by yozhao** [comment]
This is horovod generated timeline：
![nego_brod](https://user-images.githubusercontent.com/2783399/80698778-862aa980-8b0d-11ea-829f-d379ee048478.jpg)

NEGOTIATE_BROADCAST is the biggest block,  seems this is initialization work?

After zoom in, the timeline is:
![zoom_hvd](https://user-images.githubusercontent.com/2783399/80699265-41534280-8b0e-11ea-91ad-285e6baa83eb.jpg)

BROADCAST WAIT_FOR_OTHER_TENSOR_DATA  NCCL_BCAST are very short, 0.0x ms like.
I can't get a clue from this timeline.
